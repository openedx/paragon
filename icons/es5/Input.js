function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

import * as React from "react";

function SvgInput(props) {
  return /*#__PURE__*/React.createElement("svg", _extends({
    xmlns: "http://www.w3.org/2000/svg",
    width: 24,
    height: 24,
    viewBox: "0 0 24 24",
    fill: "none"
  }, props), /*#__PURE__*/React.createElement("path", {
    d: "M21 3.01H3c-1.1 0-2 .9-2 2V9h2V4.99h18v14.03H3V15H1v4.01c0 1.1.9 1.98 2 1.98h18c1.1 0 2-.88 2-1.98v-14a2 2 0 00-2-2zM11 16l4-4-4-4v3H1v2h10v3zM23 3.01H1V9h2V4.99h18v14.03H3V15H1v5.99h22V3.01zM11 16l4-4-4-4v3H1v2h10v3z",
    fill: "currentColor"
  }));
}

export default SvgInput;