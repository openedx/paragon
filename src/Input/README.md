---
title: 'Input'
type: 'component'
components:
- Input
categories:
- Forms (deprecated)
status: 'Deprecate Soon'
designStatus: 'Done'
devStatus: 'Done'
notes: |
  Replaced by Form.Control
---

A component for all user input. It is responsible for rendering select, textarea, and inputs of any type with the appropriate bootstrap class name.

Extra props supplied to Input will be passed through to the html node.

## Text

```jsx live
<Input type="text" defaultValue="Some text input" />
```

## Select

```jsx live
<Input
  type="select"
  defaultValue="Foo Bar"
  options={[
    { value: 'Foo Bar', label: 'Foo Bar' },
    { value: 'Foos Bar', label: 'Bar foo' },
    { value: 'Foo sBar', label: 'FoBaro' },
    { value: 'Foo ssBar', label: 'Farboo' },
    {
      label: 'But there is more',
      group: [
        { value: 'vFoo Bar', label: 'Foov Bar' },
        { value: 'vFoos Bar', label: 'Barv foo' },
        { value: 'vFoo sBar', label: 'FoBarov' },
        { value: 'vFoo ssBar', label: 'Farboov' },
      ],
    },
  ]}
/>
```

## Textarea

```jsx live
<Input
  type="textarea"
  defaultValue="Hammock semiotics pok pok jianbing venmo, crucifix taiyaki stumptown irony ennui knausgaard bitters synth slow-carb iPhone."
/>
```

## Date

```jsx live
<Input type="date" />
```

### File

```jsx live
<Input type="file" />
```

### Range

```jsx live
<Input type="range" />
```
