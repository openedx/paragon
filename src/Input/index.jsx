import React from 'react';
import classNames from 'classnames';
import PropTypes from 'prop-types';

class Input extends React.Component {
  componentDidMount() {
    if (process.env.NODE_ENV === 'development') {
      this.checkHasLabel();
    }
  }

  getHTMLTagForType() {
    const { type } = this.props;
    if (type === 'select' || type === 'textarea') { return type; }
    return 'input';
  }

  getClassNameForType() {
    switch (this.props.type) {
      case 'file':
        return 'form-control-file';
      case 'checkbox':
      case 'radio':
        return 'form-check-input';
      default:
        return 'form-control';
    }
  }

  setRef(forwardedRef) {
    // In production just return the optional forwardedRef
    if (process.env.NODE_ENV !== 'development') { return forwardedRef; }

    return (element) => {
      if (forwardedRef) { forwardedRef.current = element; } // eslint-disable-line no-param-reassign
      this.inputEl = element;
    };
  }

  checkHasLabel() {
    if (this.inputEl.labels.length > 0) { return; }
    if (this.inputEl.getAttribute('aria-label') !== null) { return; }
    if (this.inputEl.getAttribute('aria-labelledby') !== null) { return; }

    if (console) {
      // eslint-disable-next-line no-console
      console.warn('Input[a11y]: There is no associated label for this Input');
    }
  }

  renderOptions(options) {
    return options.map((option) => {
      const {
        value,
        label,
        group,
        ...attributes
      } = option;

      if (group) {
        return (
          <optgroup key={`optgroup-${label}`} label={label} {...attributes}>
            {this.renderOptions(group)}
          </optgroup>
        );
      }
      return (
        <option key={value} value={value} {...attributes}>
          {label}
        </option>
      );
    }, this);
  }

  render() {
    const {
      type,
      className,
      options,
      forwardedRef, // eslint-disable-line react/prop-types
      ...attributes // eslint-disable-line react/prop-types
    } = this.props;

    const htmlTag = this.getHTMLTagForType();
    const htmlProps = {
      className: classNames(this.getClassNameForType(), className),
      type: htmlTag === 'input' ? type : undefined,
      ...attributes,
      ref: this.setRef(forwardedRef),
    };
    const htmlChildren = type === 'select' ? this.renderOptions(options) : null;

    return React.createElement(htmlTag, htmlProps, htmlChildren);
  }
}

Input.propTypes = {
  /** specifies the type of component.
   * One of select, textarea, or any valid type for an html input tag. */
  type: PropTypes.oneOf([
    'textarea',
    'select',
    'checkbox',
    'color',
    'date',
    'datetime',
    'datetime-local',
    'email',
    'file',
    'hidden',
    'image',
    'month',
    'number',
    'password',
    'radio',
    'range',
    'reset',
    'search',
    'submit',
    'tel',
    'text',
    'time',
    'url',
    'week',
  ]).isRequired,
  /** specifies the className in addition to a bootstrap class name. */
  className: PropTypes.string,
  /** should be used to specify the options of an Input of type select */
  options: PropTypes.arrayOf(PropTypes.shape({
    label: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    disabled: PropTypes.bool,
    group: PropTypes.arrayOf(PropTypes.shape({
      label: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
      value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
      disabled: PropTypes.bool,
    })),
  })),
};

Input.defaultProps = {
  className: undefined,
  options: [],
};

// eslint-disable-next-line react/no-multi-comp
const InputWithRefForwarding = React.forwardRef((props, ref) => (
  <Input forwardedRef={ref} {...props} />
));

export default InputWithRefForwarding;
