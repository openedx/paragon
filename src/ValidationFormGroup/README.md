---
title: 'ValidationFormGroup'
type: 'component'
components:
- ValidationFormGroup
categories:
- Forms (deprecated)
status: 'Deprecate Soon'
designStatus: 'Done'
devStatus: 'Done'
notes: |
  Replaced by Form.Group
---

Handles bootstrap style field validation and handles related aria attributes.

Manages the rendering of bootstrap-style:

- Help text
- Valid and invalid feedback

For children of type input, textarea, and select:

- Appends bootstrap validation class names
- Appends aria-describedby attributes (for help text and feedback)

## basic usage

```jsx live
<ValidationFormGroup
  for="firstName"
  helpText="This is your name."
>
  <label htmlFor="firstName">First Name</label>
  <Input
    type="text"
    id="firstName"
    name="first-name"
    value="Casey"
    onChange={() => {}}
  />
</ValidationFormGroup>
```

## invalid message

```jsx live
<ValidationFormGroup
  for="firstName"
  invalid
  invalidMessage="Wrong!"
>
  <label htmlFor="firstName">First Name</label>
  <Input
    type="text"
    id="firstName"
    name="first-name"
    value="Casey"
    onChange={() => {}}
  />
</ValidationFormGroup>
```

## valid message

```jsx live
<ValidationFormGroup
  for="firstName"
  valid
  validMessage="What a nice name!"
>
  <label htmlFor="firstName">First Name</label>
  <Input
    type="text"
    id="firstName"
    name="first-name"
    value="Casey"
    onChange={() => {}}
  />
</ValidationFormGroup>
```

## with any kind of input

```jsx live
<>
<ValidationFormGroup
  for="weatherToday"
  helpText="Look out the window to see."
  valid
  validMessage="Correct!"
>
  <label htmlFor="weatherToday">How is the weather today?</label>
  <select
    className="form-control"
    id="weatherToday"
    name="weather"
    value="Sunny"
    onChange={() => {}}
  >
    <option>Sunny</option>
    <option>Cloudy</option>
    <option>Rainy</option>
    <option>Snowy</option>
  </select>
</ValidationFormGroup>
<ValidationFormGroup
  for="weatherTomorrow"
  helpText="Incoming weather."
  invalid
  invalidMessage="No good!"
>
  <Form.Control type='text' value='Cloudy' />
</ValidationFormGroup>
</>
```
