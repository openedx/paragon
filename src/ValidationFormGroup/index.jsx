import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import Input from '../Input';
import { FormControl } from '../Form';

const propTypes = {
  /** Id of the form input that the validation is for */
  for: PropTypes.string.isRequired,
  /** Additional classnames for this component */
  className: PropTypes.string,
  /** Determines if invalid styles / message will be shown */
  invalid: PropTypes.bool,
  /** Determines if invalid styles / message will be shown */
  valid: PropTypes.bool,
  /** Message to display on valid input */
  validMessage: PropTypes.node,
  /** Message to display on invalid input */
  invalidMessage: PropTypes.node,
  /** Help text for the form input */
  helpText: PropTypes.node,
  children: PropTypes.node,
};

const defaultProps = {
  invalid: undefined,
  valid: undefined,
  validMessage: undefined,
  invalidMessage: undefined,
  helpText: undefined,
  children: undefined,
  className: undefined,
};

function ValidationFormGroup(props) {
  const {
    className,
    invalidMessage,
    invalid,
    valid,
    validMessage,
    helpText,
    for: id,
    children,
  } = props;

  const renderChildren = () => React.Children.map(children, (child) => {
    // Any non-user input element should pass through unmodified
    if (['input', 'textarea', 'select', Input, FormControl].indexOf(child.type) === -1) { return child; }

    // Add validation class names and describedby values to input element
    return React.cloneElement(child, {
      className: classNames(child.props.className, {
        'is-invalid': invalid,
        'is-valid': valid,
      }),
      // This is a non-standard use of the classNames package, but it's exactly the same use case.
      'aria-describedby': classNames(child.props['aria-describedby'], {
        [`${id}-help-text`]: Boolean(helpText),
        [`${id}-invalid-feedback`]: invalid && invalidMessage,
        [`${id}-valid-feedback`]: valid && validMessage,
      }),
    });
  });

  const renderHelpText = (text) => {
    if (!text) { return null; }
    return <small id={`${id}-help-text`} className="form-text text-muted">{text}</small>;
  };

  /**
   * The red text conveys semantic emphasis using color and font weight. For WCAG 2.1, the
   * semantics need to be exposed programmatically as well. To do this, we use <strong/>
   * elements and attach the formatting classes to them.
   */
  const renderInvalidFeedback = (message) => {
    if (!message) { return null; }
    return (
      <strong
        id={`${id}-invalid-feedback`}
        className="invalid-feedback"
      >
        {message}
      </strong>
    );
  };

  const renderValidFeedback = (message) => {
    if (!message) { return null; }
    return (
      <div
        className="valid-feedback"
        id={`${id}-valid-feedback`}
      >
        {message}
      </div>
    );
  };

  return (
    <div className={classNames('form-group', className)}>
      {renderChildren()}
      {renderHelpText(helpText)}
      {renderInvalidFeedback(invalidMessage)}
      {renderValidFeedback(validMessage)}
    </div>
  );
}

ValidationFormGroup.propTypes = propTypes;
ValidationFormGroup.defaultProps = defaultProps;

export default ValidationFormGroup;
