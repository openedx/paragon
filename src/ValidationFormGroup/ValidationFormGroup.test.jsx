import React from 'react';
import renderer from 'react-test-renderer';
import { render, screen } from '@testing-library/react';

import Input from '../Input';
import { FormControl, Form } from '..';
import ValidationFormGroup from '.';

describe('ValidationFormGroup', () => {
  const labelAndInputComponents = (
    <>
      <label htmlFor="firstName">First Name</label>
      <input
        className="form-control"
        type="text"
        id="firstName"
        name="first-name"
        value="Adam"
        onChange={() => {}}
      />
    </>
  );

  it('renders', () => {
    const tree = renderer.create((
      <ValidationFormGroup
        for="firstName"
      >
        {labelAndInputComponents}
      </ValidationFormGroup>
    )).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders with help text', () => {
    const tree = renderer.create((
      <ValidationFormGroup
        for="firstName"
        helpText="This is your first name."
      >
        {labelAndInputComponents}
      </ValidationFormGroup>
    )).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders an invalid message', () => {
    const tree = renderer.create((
      <ValidationFormGroup
        for="firstName"
        helpText="This is your first name."
        invalid
        invalidMessage="This is not your name."
      >
        {labelAndInputComponents}
      </ValidationFormGroup>
    )).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders a valid message', () => {
    const tree = renderer.create((
      <ValidationFormGroup
        for="firstName"
        helpText="This is your first name."
        valid
        validMessage="What a nice name."
      >
        {labelAndInputComponents}
      </ValidationFormGroup>
    )).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders an invalid message with a select input', () => {
    const tree = renderer.create((
      <ValidationFormGroup
        for="weatherToday"
        helpText="Look out the window to see."
        valid
        validMessage="Correct!"
      >
        <label htmlFor="weatherToday">How is the weather today?</label>
        <select id="weatherToday" name="weather" value="Sunny" onChange={() => {}}>
          <option>Sunny</option>
          <option>Cloudy</option>
          <option>Rainy</option>
          <option>Snowy</option>
        </select>
      </ValidationFormGroup>
    )).toJSON();
    expect(tree).toMatchSnapshot();
  });
  [
    { Component: Input, name: 'Input' },
    { Component: FormControl, name: 'FormControl' },
    { Component: Form.Control, name: 'Form.Control' },
  ].forEach(({ Component, name }) => {
    const formControlId = 'bestForm';
    const formControlHelp = 'So helpful';
    const validMessage = 'Valid feedback';
    const invalidMessage = 'Exterminate, Exterminate';

    it(`renders a ${name} child with the correct aria attributes and helptext`, () => {
      render(
        <ValidationFormGroup
          for={formControlId}
          helpText={formControlHelp}
          validMessage={validMessage}
          valid
        >
          <Component id={formControlId} />
        </ValidationFormGroup>,
      );
      const component = screen.getByRole('textbox');
      expect(component.getAttribute('aria-describedby')).toBe(`${formControlId}-help-text ${formControlId}-valid-feedback`);
      expect(screen.getByText(formControlHelp)).toBeTruthy();
    });
    it(`renders a ${name} child with the correct valid message`, () => {
      render(
        <ValidationFormGroup
          for={formControlId}
          helpText={formControlHelp}
          validMessage={validMessage}
          valid
        >
          <Component id={formControlId} />
        </ValidationFormGroup>,
      );
      expect(screen.getByText(validMessage)).toBeTruthy();
    });
    it(`renders a ${name} child with the correct invalid message`, () => {
      render(
        <ValidationFormGroup
          for={formControlId}
          helpText={formControlHelp}
          invalidMessage={invalidMessage}
          invalid
        >
          <Component id={formControlId} />
        </ValidationFormGroup>,
      );
      expect(screen.getByText(invalidMessage)).toBeTruthy();
    });
  });
});
