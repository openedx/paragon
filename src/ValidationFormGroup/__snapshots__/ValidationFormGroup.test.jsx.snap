// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ValidationFormGroup renders 1`] = `
<div
  className="form-group"
  data-testid="validation-form-group"
>
  <label
    htmlFor="firstName"
  >
    First Name
  </label>
  <input
    className="form-control"
    id="firstName"
    name="first-name"
    onChange={[Function]}
    type="text"
    value="Adam"
  />
</div>
`;

exports[`ValidationFormGroup renders a valid message 1`] = `
<div
  className="form-group"
  data-testid="validation-form-group"
>
  <label
    htmlFor="firstName"
  >
    First Name
  </label>
  <input
    className="form-control"
    id="firstName"
    name="first-name"
    onChange={[Function]}
    type="text"
    value="Adam"
  />
  <small
    className="form-text text-muted"
    id="firstName-help-text"
  >
    This is your first name.
  </small>
  <div
    className="valid-feedback"
    id="firstName-valid-feedback"
  >
    What a nice name.
  </div>
</div>
`;

exports[`ValidationFormGroup renders an invalid message 1`] = `
<div
  className="form-group"
  data-testid="validation-form-group"
>
  <label
    htmlFor="firstName"
  >
    First Name
  </label>
  <input
    className="form-control"
    id="firstName"
    name="first-name"
    onChange={[Function]}
    type="text"
    value="Adam"
  />
  <small
    className="form-text text-muted"
    id="firstName-help-text"
  >
    This is your first name.
  </small>
  <strong
    className="invalid-feedback"
    id="firstName-invalid-feedback"
  >
    This is not your name.
  </strong>
</div>
`;

exports[`ValidationFormGroup renders an invalid message with a select input 1`] = `
<div
  className="form-group"
  data-testid="validation-form-group"
>
  <label
    htmlFor="weatherToday"
  >
    How is the weather today?
  </label>
  <select
    aria-describedby="weatherToday-help-text weatherToday-valid-feedback"
    className="is-valid"
    id="weatherToday"
    name="weather"
    onChange={[Function]}
    value="Sunny"
  >
    <option>
      Sunny
    </option>
    <option>
      Cloudy
    </option>
    <option>
      Rainy
    </option>
    <option>
      Snowy
    </option>
  </select>
  <small
    className="form-text text-muted"
    id="weatherToday-help-text"
  >
    Look out the window to see.
  </small>
  <div
    className="valid-feedback"
    id="weatherToday-valid-feedback"
  >
    Correct!
  </div>
</div>
`;

exports[`ValidationFormGroup renders with help text 1`] = `
<div
  className="form-group"
  data-testid="validation-form-group"
>
  <label
    htmlFor="firstName"
  >
    First Name
  </label>
  <input
    className="form-control"
    id="firstName"
    name="first-name"
    onChange={[Function]}
    type="text"
    value="Adam"
  />
  <small
    className="form-text text-muted"
    id="firstName-help-text"
  >
    This is your first name.
  </small>
</div>
`;
